dali/operators/generic/histogram/histogram.cc

```diff
@@ -0,0 +1,685 @@
+    auto sh_ranges = dim_ranges.shape();
+    DALI_ENFORCE(sh_ranges.sample_dim(),
+                 make_string("Histogram bin ranges for ", r,
+                               " dimension should be one-dimensional array"));
+
+    DALI_ENFORCE(
+      sh_ranges.num_elements() == 2 * nsamples,
+        make_string("Bin ranges for uniform histogram should consist of lower and upper bound",
+                    dim_ranges.num_samples()));
+
+    for (int i = 0; i < nsamples; ++i) {
+      batch_ranges_.push_back(GetFlattenedRanges(i, ws));
+    }
+  }
+  return hist_dim;
+}
+
+int HistogramCPU::ValidateRangeArguments(const workspace_t<CPUBackend> &ws, int num_samples) {
+  if (uniform_) {
+    return ValidateUniformRangeArguments(ws, num_samples);
+  } else {
+    return ValidateNonUniformRangeArguments(ws, num_samples);
+  }
+  // TODO: validate that bin ranges grow monoticaly
+  // currently its checked later on by OpenCV
+}
+
+void HistogramCPU::RunImpl(workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.Input<CPUBackend>(0);
+  auto &output = ws.Output<CPUBackend>(0);
+
+  int nsamples = input.shape().num_samples();
+  auto &thread_pool = ws.GetThreadPool();
+  std::vector<int> all_channels;
+
+  for (int j = 0; j < hist_dim_; ++j) {
+    all_channels.push_back(j);
+  }
+
+  TYPE_SWITCH(input.type(), type2id, Type, (uint8_t, uint16_t, float), (
+  {
+    auto in_view = view<const Type>(input);
+
+    if (is_identity_) {
+      auto out_view_id = view<Type>(output);
+      run_identity<Type>(thread_pool, in_view, out_view_id);
+       return;
+    }
+
+    auto out_view = view<float>(output);
+    TensorListView<StorageCPU, const Type, DynamicDimensions> transposed_in_view;
+
+    bool needs_transpose = NeedsTranspose();
+    if (needs_transpose) {
+      transpose_mem_.template Reserve<mm::memory_kind::host>(
+        in_view.num_elements() * sizeof(Type));
+      auto scratch = transpose_mem_.GetScratchpad();
+
+      transposed_in_view =
+      transpose_view(thread_pool, scratch, in_view, axes_order_);
+    }
+
+    auto splited_in_views = reinterpret<Type>(
+      needs_transpose ? transposed_in_view : in_view, splited_input_shapes_);
+    auto splited_out_views = reinterpret<float>(out_view, splited_output_shapes_);
+
+    assert(splited_in_views.num_samples() == splited_out_views.num_samples());
+    assert(split_mapping_.size() == size_t(splited_in_views.num_samples()));
+
+    for (int i = 0; i < splited_in_views.num_samples(); ++i) {
+      thread_pool.AddWork([&, i](int thread_id) {
+        SmallVector<int, 2> in_sizes;
+
+        auto in_shape_span = splited_in_views.tensor_shape_span(i);
+        in_sizes.push_back(in_shape_span[0]);  // FIXME, volume?
+
+        int in_type = CVMatType<Type>::get(hist_dim_);
+
+        std::vector<cv::Mat> images = {
+          cv::Mat(1, in_sizes.data(), in_type, splited_in_views[i].data)};
+
+        cv::InputArray input_mat(images);
+
```

Comment:
I don't think you need an `cv::InputArray`. You can pass `cv::Mat` to `cv::calcHist`